fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanApplicationWritePlatformServiceJpaRepositoryImpl.java

```diff
@@ -794,19 +794,25 @@ public CommandProcessingResult rejectApplication(final Long loanId, final JsonCo
 
         final AppUser currentUser = getAppUserIfPresent();
 
-        this.loanApplicationTransitionApiJsonValidator.validateRejection(command.json());
+        // validate before loan assembling
+        loanApplicationTransitionValidator.validateRejection(command.json());
 
+        // retrieve/assemble loan
         final Loan loan = retrieveLoanBy(loanId);
 
+        // check client or group
         checkClientOrGroupActive(loan);
 
+        // check for mandatory entities
         entityDatatableChecksWritePlatformService.runTheCheckForProduct(loanId, EntityTables.LOAN.getName(),
                 StatusEnum.REJECTED.getCode().longValue(), EntityTables.LOAN.getForeignKeyColumnNameOnDatatable(), loan.productId());
 
+        // validate loan rejection
+        loanApplicationTransitionValidator.validateRejection(loan, defaultLoanLifecycleStateMachine, command);
+
         final Map<String, Object> changes = loan.loanApplicationRejection(currentUser, command, defaultLoanLifecycleStateMachine);
```

Comment:
Please extract out logic from loan.loanApplicationRejection method into this service.

Reply:
done